### PR TITLE
[Merged by Bors] - Consistently use `PI` to specify angles in examples.

### PR DESCRIPTION
**Pull request by devil-ira**
Examples inconsistently use either `TAU`, `PI`, `FRAC_PI_2` or `FRAC_PI_4`.
Often in odd ways and without `use`ing the constants, making it difficult to parse.

 * Use `PI` to specify angles.
 * General code-quality improvements.
 * Fix borked `hierarchy` example.

### CODE REVIEW

File: examples/3d/lighting.rs [hunk 1]
```diff
@@ -34,7 +36,7 @@ fn setup(
 
     // left wall
     let mut transform = Transform::from_xyz(2.5, 2.5, 0.0);
-    transform.rotate_z(std::f32::consts::FRAC_PI_2);
+    transform.rotate_z(TAU / 4.);
```
Comment (bjorn3) [inline examples/3d/lighting.rs:39]: This has less accuracy than FRAC_PI_2.

Comment (devil-ira) [reply examples/3d/lighting.rs:39]: I don't think the loss of precision is relevant to any of our current examples,
but I wanted to investigate the difference in precision so i tried this:
```rust
assert_eq!(
    std::f64::consts::TAU / 4.,
    std::f64::consts::FRAC_PI_2
);
```
This assertion does not fail however. Is this test too naive?

File: examples/2d/mesh2d_manual.rs [hunk 2]
```diff
@@ -62,12 +64,12 @@ fn star(
     // These vertices are specified in 3D space.
     let mut v_pos = vec![[0.0, 0.0, 0.0]];
     for i in 0..10 {
-        // Angle of each vertex is 1/10 of TAU, plus PI/2 for positioning vertex 0
-        let a = std::f32::consts::FRAC_PI_2 - i as f32 * std::f32::consts::TAU / 10.0;
-        // Radius of internal vertices (2, 4, 6, 8, 10) is 100, it's 200 for external
+        // The angle between each vertex is 1/10 of a full rotation.
+        let a = i as f32 * TAU / 10.0;
```
Comment (DJMcNab) [inline examples/2d/mesh2d_manual.rs:68]: I'd be tempted to use degrees here, and then convert to radians as needed.

File: examples/3d/lighting.rs (hunk 1 above)
Comment (DJMcNab) [reply examples/3d/lighting.rs:39]: I'd compare the `to_bits` outcome to be safe

**Review by DJMcNab — COMMENTED** [review]
So in theory I think this makes sense, however there are a few things which make me hesitant. In my experience, education about how to think about radians is all in terms of Pi, not Tau. Therefore, this runs into issues of causing challenges for those with an intuition for radians in the standard forms.

Basically, none of this code is self describing to an audience without a formal education in relatively advanced maths, and familiarity with Tau. I think this is unnecessary. There are two paths forward I can see.

One is to recognise that the name `Tau` is not at all evocative of its meaning. Therefore, we can add `const FULL_ROTATION_RADIANS: f64 = TAU;` in `bevy_utils`. This has the issue of being a non-standard constant, but the name should be descriptive enough to make reasoning about it much easier.

The second is to give up on radians entirely, trust llvm to optimise things properly and use degrees, converting as needed. The main issue I can see there is that there is (currently :eyes:) no type system support for that, so it's potentially easy to accidentally pass e.g. 180 radians to a function, causing very weird bugs.

**Comment by Carter0** [comment]
> 



> So in theory I think this makes sense, however there are a few things which make me hesitant. In my experience, education about how to think about radians is all in terms of Pi, not Tau. Therefore, this runs into issues of causing challenges for those with an intuition for radians in the standard forms.
> 
> Basically, none of this code is self describing to an audience without a formal education in relatively advanced maths, and familiarity with Tau. I think this is unnecessary. There are two paths forward I can see.
> 
> One is to recognise that the name `Tau` is not at all evocative of its meaning. Therefore, we can add `const FULL_ROTATION_RADIANS: f64 = TAU;` in `bevy_utils`. This has the issue of being a non-standard constant, but the name should be descriptive enough to make reasoning about it much easier.
> 
> The second is to give up on radians entirely, trust llvm to optimise things properly and use degrees, converting as needed. The main issue I can see there is that there is (currently 👀) no type system support for that, so it's potentially easy to accidentally pass e.g. 180 radians to a function, causing very weird bugs.

I think TAU is a lot easier to understand than PI once you get the hang of it. And almost all game math libraries use radians so I don't really think it makes sense to use degrees anywhere.

File: examples/3d/lighting.rs (hunk 1 above)
Comment (devil-ira) [reply examples/3d/lighting.rs:39]: Same result with `to_bits`.

**Comment by devil-ira** [comment]
> Basically, none of this code is self describing to an audience without a formal education in relatively advanced maths

This makes TAU seem like something tricky when it is rather straightforward.
If someone knows PI then they know that PI * 2 is a full rotation.
The missing piece would be knowing that TAU == PI * 2 which is easy enough to memorize.

And for people new to radians altogether. Teaching that TAU is the same as 360 degrees is more palatable than PI being 180 degrees and having to think in half-rotations.

Reading `TAU / 4.` it's quick to see that that is a `1/4` of a turn.
Just like `TAU * 2. * time.delta_seconds()` is something being rotated `2` times per second.
Comparing that to PI:
`PI / 2.` or `FRAC_PI_2` is 1/4 of a turn.
`PI * 4 * time.delta_seconds()` is something being rotated `2` times per second.
You end up having to multiply or divide by `2` in your head.

> familiarity with Tau

Fair. Can't understand what you don't know, and schools don't seem to teach it.
It *is* very simple though.

> One is to recognise that the name Tau is not at all evocative of its meaning.

Neither is PI, but of course more people already know PI.
I'd rather teach people what TAU means through comments in the examples.
`FULL_ROTATION_RADIANS` would need an `f32` version as well.
It's also even more verbose than `to_radians`.

> The second is to give up on radians entirely

:sweat_smile: kind of difficult when every math lib uses radians. Trust me. I'd replace everything with degrees if I could, hah.
It is probably worthwhile to switch to using `to_radians` in some examples though,
but it's too verbose to use everywhere and radians are unavoidable. `180f32.to_radians()` vs `TAU / 2.`.

TL;DR
Thinking of angles in terms of full rotations is intuitive, but then PI requires dividing/multiplying by two in your head every time you read or write an angle.
TAU is a very simple constant. A full rotation in radians. People more familiar with PI would still pick it up very quickly.
I expect people new to using radians have an easier time learning if they are taught to use TAU as a full rotation rather than PI as a half rotation.

Using degrees would be nice but since we need to use radians in the end it introduces verbose conversions in the middle of the actual math.

I'm going to try to use `to_radians` in some of the examples and see how that goes.

**Comment by cart** [comment]
> I think TAU is a lot easier to understand than PI once you get the hang of it. 

Thats the problem. Examples are intended to be minimal / quick / easy to understand / etc and are often peoples' first exposure to the engine.

The percentage of programmers that know what PI represents off the top of their head is probably close to 98%. The percentage of programmers that know what TAU represents off the top of their heard is probably closer to 5%.

It doesn't matter that it is a "simple" concept (or marginally more "natural" once you understand it). The vast majority of people think about radians in terms of PI. Using TAU in our examples adds one more (unnecessary) learning hurdle to the Bevy onboarding process, which is already full of necessary learning hurdles.

"Optimizing" this part of peoples' workflows doesn't seem worth the price of admission, especially in this part of our learning material.

**Comment by rparrett** [comment]
I don't have an opinion about the TAU/PI debate, but I appreciated all the little bits of incidental cleanup in this PR, especially importing (your favorite constant). I checked a few of the examples where the changes were less trivial and they checked out.

**Review by afonsolage — APPROVED** [review]
Besides the [TAU vs PI](https://xkcd.com/1292/) discussion, I think the reworks is pretty good.

I'm in favor of `TAU`, since I think it's easier to understand, but the majority of newcomers won't, most of them still struggle to understand radians vs degree, so maybe we should add on `bevy_utils` some consts like `DEGREE_360`, `DEGREE_180`, `DEGREE_90`, `DEGREE_45` and accept that examples will be better written in degrees, since anyone who knows `PI` or `TAU` will understand what is a `transform.rotate_z(DEGREE_90)`

**Comment by devil-ira** [comment]
> "Optimizing" this part of peoples' workflows doesn't seem worth the price of admission, especially in this part of our learning material.

You're right. The most important thing is the examples being easy to start out with.
I was thinking that since there is already quite a few things to learn adding one more wouldn't hurt, but I realize now that is not the right way to think about it.

> consts like DEGREE_360, DEGREE_180, DEGREE_90, DEGREE_45

`DEGREE_360` *is* 360 degrees... but in radians, that might be confusing.

<sub>PS: A *lot* of the discussion around PI/TAU is *extremely* cringe.
The only argument I'll make is that TAU avoids some mental math when reading/writing for me personally.
If you want a good laugh, go read the PI/TAU manifestos. They're so bad haha.

**Review by cart — APPROVED** [review]
This looks good to me. My last holdout was the precision of `PI / 2.` and `PI / 4.` vs `FRAC_PI_2` and `FRAC_PI_4`. (a previous comment already showed this wasn't a problem for TAU, but that doesn't necessarily make it true for PI).

They are equal!

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f3c460954233270951053011f30d5fcb

Under those circumstances, I agree with the choice to use the one `PI` constant for this. Simpler and more flexible.
